catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsLoader.java

```diff
@@ -0,0 +1,223 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import org.apache.commons.io.IOUtils;
+import org.boon.Boon;
+import org.codice.ddf.catalog.ui.forms.data.QueryTemplateMetacardImpl;
+import org.codice.ddf.catalog.ui.forms.data.ResultTemplateMetacardImpl;
+import org.codice.ddf.configuration.AbsolutePathResolver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Loads initial system template configuration from the file system so template defaults in
+ * distributions can vary independently.
+ */
+public class SearchFormsLoader implements Supplier<List<Metacard>> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SearchFormsLoader.class);
+
+  private static final File DEFAULT_FORMS_DIRECTORY =
+      new File(new AbsolutePathResolver("etc/forms").getPath());
+
+  private static final String FORMS_FILE_NAME = "forms.json";
+
+  private static final String RESULTS_FILE_NAME = "results.json";
+
+  private final File configDirectory;
+
+  public static Supplier<List<Metacard>> config() {
+    return new SearchFormsLoader();
+  }
+
+  public SearchFormsLoader() {
+    this(DEFAULT_FORMS_DIRECTORY);
+  }
+
+  public SearchFormsLoader(File configDirectory) {
+    this.configDirectory = configDirectory;
+  }
+
+  @Override
+  public List<Metacard> get() {
+    if (!configDirectory.exists()) {
+      LOGGER.debug("Could not locate forms directory");
+      return Collections.emptyList();
+    }
+
+    // What's our policy here? Do we want the system to blowup / make noise if this isn't possible?
+    // Especially due to the security manager?
+    if (!configDirectory.canRead()) {
+      LOGGER.debug("Forms directory exists but could not be read");
+      return Collections.emptyList();
+    }
+
+    File formsFile = configDirectory.toPath().resolve(FORMS_FILE_NAME).toFile();
+    File resultsFile = configDirectory.toPath().resolve(RESULTS_FILE_NAME).toFile();
+
+    return Stream.concat(
+            loadFile(formsFile, this::formMapper), loadFile(resultsFile, this::resultsMapper))
+        .collect(Collectors.toList());
+  }
+
+  /** Read the provided JSON file and return a stream of metacards. */
+  private Stream<Metacard> loadFile(File file, Function<? super Map, Metacard> mapper) {
+    if (!file.exists()) {
+      LOGGER.debug("Could not locate {}", file.getName());
+      return Stream.empty();
+    }
+
+    String payload = getFileContent(file);
+    if (payload == null) {
+      LOGGER.debug("Problem reading {}", file.getName());
+      return Stream.empty();
+    }
+
+    try {
+      return Stream.of(Boon.fromJson(payload))
+          .map(List.class::cast)
+          .flatMap(List::stream)
+          .filter(Map.class::isInstance)
+          .map(Map.class::cast)
+          .map(mapper)
+          .filter(Objects::nonNull)
+          // Invoke a terminal operation so the exception can be handled here
+          .collect(Collectors.toList())
```

Comment:
Static analysis (and IntelliJ as well, I believe) actually complain that extracting the variable is a sub-optimal choice when the entire thing can be streamed so cleanly. But style aside...

Performing the `List<...>` cast will negate all of the per-element error handling logic, and as a result, even template entries that are correct will be ignored. 

I think that complete failure should only occur when the entire configuration structure is wrong, such that no individual elements can be validated. The catalog should still be initialized with correct elements even when others are wrong. Do we think this will take the user off guard?

Reply:
I found a way to rework this to make things clearer.